dbms/src/Functions/FunctionDateOrDateTimeAddInterval.h

```diff
@@ -286,20 +377,60 @@ class FunctionDateOrDateTimeAddInterval : public IFunction
                     ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
         }
 
-        if (WhichDataType(arguments[0].type).isDate())
+        switch (arguments[0].type->getTypeId())
         {
-            if (std::is_same_v<decltype(Transform::execute(DataTypeDate::FieldType(), 0, std::declval<DateLUTImpl>())), UInt16>)
-                return std::make_shared<DataTypeDate>();
-            else
-                return std::make_shared<DataTypeDateTime>(extractTimeZoneNameFromFunctionArguments(arguments, 2, 0));
+            case TypeIndex::Date:
+                return resolveReturnType<DataTypeDate>(arguments);
+            case TypeIndex::DateTime:
+                return resolveReturnType<DataTypeDateTime>(arguments);
+            case TypeIndex::DateTime64:
+                return resolveReturnType<DataTypeDateTime64>(arguments);
+            default:
+            {
+                // TODO (vnemkov): quick and dirty way to check, remove before merging.
+                assert(false);
+                throw Exception("Invalid type of 1st argument of function " + getName() + ": "
+                    + arguments[0].type->getName() + ", expected: Date, DateTime or DateTime64.",
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            }
+        }
+    }
+
+    // Helper templates to deduce return type based on argument type, since some overloads may promote or denote types, e.g. addSeconds(Date, 1) => DateTime
+    template <typename FieldType>
+    using TransformExecuteReturnType = decltype(std::declval<Transform>().execute(FieldType(), 0, std::declval<DateLUTImpl>()));
+
+    // Deduces RETURN DataType from INTPUT DataType, based on return type of Transform{}.execute(INPUT_TYPE, UInt64, DateLUTImpl).
+    // e.g. for Transform-type that has execute()-overload with 'UInt16' input and 'UInt32' return,
+    // argument type is expected to be 'Date', and result type is deduced to be 'DateTime'.
+    template <typename FromDataType>
+    using TransformResultDataType = typename date_and_time_type_details::ResultDataTypeMap<TransformExecuteReturnType<typename FromDataType::FieldType>>::ResultDataType;
+
+    template <typename FromDataType>
+    DataTypePtr resolveReturnType(const ColumnsWithTypeAndName & arguments) const
+    {
+        using ResultDataType = TransformResultDataType<FromDataType>;
+
+        if constexpr (std::is_same_v<ResultDataType, DataTypeDate>)
+            return std::make_shared<DataTypeDate>();
+        else if constexpr (std::is_same_v<ResultDataType, DataTypeDateTime>)
+        {
+            return std::make_shared<DataTypeDateTime>(extractTimeZoneNameFromFunctionArguments(arguments, 2, 0));
+        }
+        else if constexpr (std::is_same_v<ResultDataType, DataTypeDateTime64>)
+        {
+            // TODO (vnemkov): what if there is an overload of Transform::execute() that returns DateTime64 from DateTime or Date ?
```

Comment:
I think it's ok to assume that it's not possible now. If it is, it may be fixed later.

Reply:
Yep, there seems to be no transform-callable that does that now